### ISSUE TITLE
Return null from readQuery when data is not in store

### ISSUE THREAD

**Issue by MaxDesiatov** [apollographql/apollo-feature-requests#1]
**How to reproduce the issue:**

```js
const data = proxy.readQuery({query: gql`{ notInTheStore }`})
```

**Intended outcome:**

Returns null

**Actual outcome:**

Throws an error: `ERROR Error: Uncaught (in promise): Error: Can't find field notInTheStore on object (ROOT_QUERY) undefined.`

Migrated from:
apollographql/apollo-client#2007
apollographql/apollo-client#1542

**Comment by bbugh** [comment]
Longer discussion of this on a closed task (it is still a problem) https://github.com/apollographql/apollo-client/issues/1542

**Comment by lorensr** [comment]
Thanks @bbugh. Updated title—it looks like returning null is the favored behavior. 

I'm for labeling `help-wanted`

**Comment by bbugh** [comment]
The consensus does seem to be to return `null`, but I think perhaps `undefined` is more correct for this use case. It's reasonable to expect that a valid `readQuery` would return a `null` value if the query is found *and* the value is actually null. I expect this is why the `throw` is there in the first place. Thus, I think the return value when a query isn't found should be `undefined`. Here's some silly code to demonstrate:

```javascript
const query = gql`{ thing { name } }`
const data = { thing: { name: null } }

// Write a null value to the cache
store.writeQuery({ query, data })

// found query on readQuery, so it should return the value
store.readQuery(query)
=> null  // this is *actually* null

// proposal for null: doesn't make sense to return null because the query doesn't exist
// no way to tell if this is actually a real value OR the query failed
store.readQuery(gql`{ missingQuery { id } }`)
=> null  

// proposal for undefined: query doesn't exist so it's not defined
store.readQuery(gql`{ missingQuery { id } }`)
=> undefined // correct - it's not actually defined
```

**Comment by nicky-lenaers** [comment]
Despite this being quite an annoying bug that forces us into catching the error returned from `readQuery`, I've created a minimal workaround to ease the pain and centrally catch the error:

```ts
import { Injectable } from '@angular/core';
import { DataProxy } from 'apollo-cache';

@Injectable({
  providedIn: 'root'
})
export class GraphQLService {

  public readQuery<T>(proxy: DataProxy, query: any): T | undefined {
    try {
      const data = proxy.readQuery<T>({ query });
      return data;
    } catch {
      return undefined;
    }
  }
}
```

You can consume this service anywhere and use it to "safely" use `readQuery`. _For now_.

**Comment by mcdougal** [comment]
I have also been using a workaround (albeit a pretty ugly one). The code below has allowed me to call `readQuery` the normal way but it will trap for the error and return `undefined`. My hope is if the behavior suggested in this issue ever makes it in to Apollo I can just remove these handful of lines without touching the rest of the codebase.

```js
  const cache = new InMemoryCache();

  // TODO: Monkey-patching in a fix for an open issue suggesting that
  // `readQuery` should return null or undefined if the query is not yet in the
  // cache: https://github.com/apollographql/apollo-feature-requests/issues/1
  cache.originalReadQuery = cache.readQuery;
  cache.readQuery = (...args) => {
    try {
      return cache.originalReadQuery(...args);
    } catch (err) {
      return undefined;
    }
  };

  return new ApolloClient({
    link,
    cache,
  });
```

**Comment by seanharr11** [comment]
I'd like to contend that this is a bug and **not** a feature-request.

The [type definition](https://github.com/apollographql/apollo-client/blob/78f7c7976c338778ea13f5d3ef0679326de3f9f2/packages/apollo-cache/src/types/DataProxy.ts#L89) of `readQuery` suggests that the method can return `null`, which makes it even more unexpected that we get an exception. If I query a resource in a datastore (i.e. `SELECT * FROM foo`), by default I get an empty collection back, not a `DatabaseError`.

Further, if the official [apollo docs](https://github.com/apollographql/apollo-client/blob/78f7c7976c338778ea13f5d3ef0679326de3f9f2/packages/apollo-cache/src/types/DataProxy.ts#L89) suggest that the `update()` method is the:
> "...recommended way of updating the cache after a query"

...then we really should have expected, intuitive & tight behavior here. This bug opens up a whole new class of runtime errors that won't be found until production.

Speaking from experience, these occur quite frequently as new features are rolled out inproduction (_lots of empty collections throughout the app's state!_). When a user creates a new resource in a collection, but has never "fetched that collection via an apollo query", when we try to `readQuery()` we get an exception & app crash.

This feels like a good candidate to be supported in v3.0. It will break interface of those who rely on exception to catch this error, but it's now or never with the pending major release!

**Comment by heymanhn** [comment]
@mcdougal thanks for that monkey-patching solution you suggested. Seems most prudent and the right place to catch the errors or invariant violations. 

+1 on having the Apollo team address this in v3.0.

**Comment by eltonio450** [comment]
this is clearly a non expected behavior (and very difficult to discover, it happens only in very particular cases)

I am not sure to understand in what circumstances readQuery returns null

**Comment by tuff** [comment]
This is a bug and not a feature request.

**Comment by heymanhn** [comment]
Agreed. Where is the right place to file this then?

**Comment by dylanwulf** [comment]
Actually this is documented behavior, so it's not a bug.

> cache.readQuery will throw an error if the data isn't in the cache, so we need to provide an initial state
https://www.apollographql.com/docs/react/data/local-state/#writequery-and-readquery

**Comment by heymanhn** [comment]
The reason this issue is here and still open is because many of us do consider this a bug, and are asking for Apollo to reconsider the current documented behavior.

**Comment by jannes-io** [comment]
Type for readQuery already has null, so anyone using TypeScript should have backwards compatibility covered.
```ts
readQuery<QueryType, TVariables = any>(
    options: DataProxy.Query<TVariables>,
    optimistic?: boolean,
): QueryType | null;
```

Any update on this feature/bug?

**Comment by impowski** [comment]
Any ETA on this one?

**Comment by dylanwulf** [comment]
@impowski Check out the commit message here: https://github.com/apollographql/apollo-client/commit/6a4e4608bd0747f691e23942eacbb6545007d24f

Looks like it'll be included in v3

**Comment by impowski** [comment]
@dylanwulf oh, I’ve seen it but didn’t read the last part. Thanks.

**Comment by geiszla** [comment]
v3 is already out, but I'm getting errors when attempting to query non-existing fields. Is there any news on this issue?

**Comment by vfonic** [comment]
I'm just in the process of upgrading to v3 and this is one of the issues I'm concerned with the most. Any updates?

**Comment by LydiaF** [comment]
If it helps, I was only using readQuery to then use writeQuery with modified data, but I learned the hard way that cache.modify is the better way to go. So I have just avoided this issue entirely :)

**Comment by JesseZomer** [comment]
Any updates on this?

**Comment by benjamn** [comment]
@JesseZomer + others: it's finally happening! https://github.com/apollographql/apollo-client/pull/7098

Fine print: we would very much appreciate everyone's help validating these changes once they are released in a beta version of Apollo Client 3.3 (see #7015). If this turns out to be a major breaking change (🤞 ), it may have to wait for AC4, but we are hopeful that it will not be very disruptive, since it merely turns exceptions into `null` return values, and `null` was already a possible return value.

Thanks to everyone for expressing your support for this change over the past two years!
